tools/epoch-sync/src/cli.rs

```diff
@@ -0,0 +1,293 @@
+                    last_header.raw_timestamp(),
+                );
+
+                *last_block_info.epoch_id_mut() = last_header.epoch_id().clone();
+                *last_block_info.epoch_first_block_mut() = first_block_hash;
+
+                let next_epoch_first_hash = hash_to_next_hash[last_header.hash()];
+                tracing::debug!("Creating EpochSyncInfo from block {:?}", last_header);
+
+                let epoch_sync_info = chain_update.create_epoch_sync_info(
+                    &last_block_info,
+                    &next_epoch_first_hash,
+                    Some(&hash_to_prev_hash),
+                )?;
+
+                let calculated_epoch_sync_data_hash_result =
+                    epoch_sync_info.calculate_epoch_sync_data_hash();
+                let canonical_epoch_sync_data_hash_result =
+                    epoch_sync_info.get_epoch_sync_data_hash();
+
+                if let Ok(calculated_epoch_sync_data_hash) = calculated_epoch_sync_data_hash_result
+                {
+                    if let Ok(Some(canonical_epoch_sync_data_hash)) =
+                        canonical_epoch_sync_data_hash_result
+                    {
+                        if calculated_epoch_sync_data_hash == canonical_epoch_sync_data_hash {
+                            tracing::info!(
+                                "EpochSyncInfo for height {:?} OK",
+                                epoch_sync_info.epoch_info.epoch_height()
+                            );
+                            continue;
+                        }
+                    }
+                }
+                tracing::error!(
+                    "EpochSyncInfo for height {:?} ERROR {:?} {:?}",
+                    epoch_sync_info.epoch_info.epoch_height(),
+                    calculated_epoch_sync_data_hash_result,
+                    canonical_epoch_sync_data_hash_result
+                );
+                num_errors += 1;
+            } else {
+                cur_hash = hash_to_prev_hash[&cur_hash];
+            }
+        }
+        assert_eq!(num_errors, 0);
+        Ok(())
+    }
+}
+
+fn get_hash_to_prev_hash_from_block_info(
+    storage: &NodeStorage,
+) -> anyhow::Result<HashMap<CryptoHash, CryptoHash>> {
+    let mut hash_to_prev_hash = HashMap::new();
+    let store = storage.get_split_store().unwrap_or(storage.get_hot_store());
+    for result in store.iter(DBCol::BlockInfo) {
+        let (_, value) = result?;
+        let block_info =
+            BlockInfo::try_from_slice(value.as_ref()).expect("Failed to deser BlockInfo");
+        if block_info.hash() != block_info.prev_hash() {
+            hash_to_prev_hash.insert(*block_info.hash(), *block_info.prev_hash());
+        }
+    }
+    Ok(hash_to_prev_hash)
+}
+
+fn get_hash_to_prev_hash_from_block_header(
+    storage: &NodeStorage,
+) -> anyhow::Result<HashMap<CryptoHash, CryptoHash>> {
+    let mut hash_to_prev_hash = HashMap::new();
```

Comment:
As you are iterating over all headers available, I think you may encounter forks.
Are forks garbage collected in the normal garbage collection or sooner (in finalization of blocks?)?

Reply:
Even with forks mapping from `cur_hash` to `prev_hash` is unique, as there will not be two blocks with the same hash, so there will not be key collision. Value collision may happen, but it is irrelevant for `HashMap`.

Reply:
Please add this as a comment to the code.